what4/src/What4/Expr/GroundEval.hs

```diff
@@ -84,6 +85,20 @@ type family GroundValue (tp :: BaseType) where
   GroundValue (BaseArrayType idx b) = GroundArray idx b
   GroundValue (BaseStructType ctx)  = Ctx.Assignment GroundValueWrapper ctx
 
+-- | Return a ground representation of a value, if it is ground.
+asGround :: IsExpr e => e tp -> Maybe (GroundValue tp)
+asGround x =
+  case exprType x of
+    BaseBoolRepr       -> asConstantPred x
+    BaseIntegerRepr    -> asInteger x
+    BaseRealRepr       -> asRational x
+    BaseStringRepr _si -> asString x
+    BaseComplexRepr    -> asComplex x
+    BaseBVRepr _w       -> asBV x
+    BaseFloatRepr _fpp  -> asFloat x
+    BaseStructRepr _   -> asStruct x >>= traverseFC (fmap GVW . asGround)
+    BaseArrayRepr _idx _tp -> Nothing
```

Comment:
I think we can do a little better here than always returning `Nothing` for arrays. See the corresponding case in `asConcrete`:

https://github.com/GaloisInc/what4/blob/c4573936f98d1deb0423af0b489ca63b34aaf4c8/what4/src/What4/Interface.hs#L3213-L3218

This can return `Just` if the all the array elements are a constant value, which is a slight improvement. (We'd need to store the result in [`ArrayConcrete`](https://hackage-content.haskell.org/package/what4-1.7/docs/What4-Expr-GroundEval.html#v:ArrayConcrete) instead of `ConcreteArray`.)

Reply:
I think the TODO is more referring to the fact that this implementation is rather conservative and will miss concrete arrays such as `[0 :-> True; _ :-> False]`, i.e., arrays where most indices map to a constant, but with a finite number of indices mapping to a different value. I suppose it wouldn't hurt to clarify the TODO to reflect this, although I am not sure if we can necessarily do better given the current `what4` interface.